File: articles/genomics-sequence-analysis-biological-data/c/genomics_sequence_kernel.c
/*
 * Compact genomics sequence statistics kernel in C.
 */

#include <ctype.h>
#include <stdio.h>
#include <string.h>

double gc_content(const char *sequence) {
    double valid = 0.0;
    double gc = 0.0;

    for (size_t i = 0; i < strlen(sequence); i++) {
        char base = toupper(sequence[i]);

        if (base == 'A' || base == 'T') {
            valid += 1.0;
        } else if (base == 'G' || base == 'C') {
            valid += 1.0;
            gc += 1.0;
        }
    }

    if (valid == 0.0) return 0.0;
    return gc / valid;
}

int ambiguous_count(const char *sequence) {
    int count = 0;

    for (size_t i = 0; i < strlen(sequence); i++) {
        char base = toupper(sequence[i]);

        if (!(base == 'A' || base == 'C' || base == 'G' || base == 'T')) {
            count++;
        }
    }

    return count;
}

int hamming_distance(const char *a, const char *b) {
    int distance = 0;

    if (strlen(a) != strlen(b)) {
        return -1;
    }

    for (size_t i = 0; i < strlen(a); i++) {
        if (toupper(a[i]) != toupper(b[i])) {
            distance++;
        }
    }

    return distance;
}

int main(void) {
    const char *sequence_a = "ATGCGCGTAATTAACCGGTTACCGTAGCTA";
    const char *sequence_b = "ATGCGCGTAATTAACCGGTTACCGTAACTA";

    printf("sequence_length=%lu\n", strlen(sequence_a));
    printf("gc_content=%.6f\n", gc_content(sequence_a));
    printf("ambiguous_bases=%d\n", ambiguous_count(sequence_a));
    printf("hamming_distance=%d\n", hamming_distance(sequence_a, sequence_b));

    return 0;
}

File: articles/genomics-sequence-analysis-biological-data/julia/genomics_sequence_kernel.jl
# Genomics sequence kernel in Julia.

function gc_content(sequence)
    seq = uppercase(sequence)
    valid = [base for base in seq if base in ['A', 'C', 'G', 'T']]
    if length(valid) == 0
        return NaN
    end
    gc = count(base -> base == 'G' || base == 'C', valid)
    return gc / length(valid)
end

function ambiguous_count(sequence)
    seq = uppercase(sequence)
    return count(base -> !(base in ['A', 'C', 'G', 'T']), seq)
end

function hamming_distance(a, b)
    if length(a) != length(b)
        error("Sequences must have equal length.")
    end
    return sum(collect(uppercase(a)) .!= collect(uppercase(b)))
end

sequence_a = "ATGCGCGTAATTAACCGGTTACCGTAGCTA"
sequence_b = "ATGCGCGTAATTAACCGGTTACCGTAACTA"

println("length=", length(sequence_a))
println("gc_content=", round(gc_content(sequence_a), digits=5))
println("ambiguous_bases=", ambiguous_count(sequence_a))
println("hamming_distance=", hamming_distance(sequence_a, sequence_b))
